cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -261,58 +331,75 @@ protected void rebalanceForBroker(Broker broker,
     boolean unbalanced = false;
     if (requireLessLoad) {
-      unbalanced = rebalanceByMovingLoadOut(broker, clusterModel, optimizedGoals, REPLICA_MOVEMENT, excludedTopics);
+      if (rebalanceByMovingLoadOut(broker, clusterModel, optimizedGoals, REPLICA_MOVEMENT, excludedTopics)) {
+        unbalanced = rebalanceBySwappingLoadOut(broker, clusterModel, optimizedGoals, excludedTopics);
+      }
     } else if (requireMoreLoad) {
-      unbalanced = rebalanceByMovingLoadIn(broker, clusterModel, optimizedGoals, REPLICA_MOVEMENT, excludedTopics);
+      if (rebalanceByMovingLoadIn(broker, clusterModel, optimizedGoals, REPLICA_MOVEMENT, excludedTopics)) {
+        unbalanced = rebalanceBySwappingLoadIn(broker, clusterModel, optimizedGoals, excludedTopics);
+      }
     }
 
     if (!unbalanced) {
       LOG.debug("Successfully balanced {} for broker {} by moving leaders and replicas.", resource(), broker.id());
     }
-
   }
 
-  protected boolean rebalanceByMovingLoadIn(Broker broker,
-                                            ClusterModel clusterModel,
-                                            Set<Goal> optimizedGoals,
-                                            ActionType actionType,
-                                            Set<String> excludedTopics)
-      throws AnalysisInputException, ModelInputException {
-    PriorityQueue<Broker> eligibleBrokers = new PriorityQueue<>(
-        (b1, b2) -> Double.compare(utilizationPercentage(b2), utilizationPercentage(b1)));
-    double clusterUtilizationPercentage =
-        clusterModel.load().expectedUtilizationFor(resource()) / clusterModel.capacityFor(resource());
-    clusterModel.healthyBrokers().forEach(b -> {
-      if (utilizationPercentage(b) > clusterUtilizationPercentage) {
-        eligibleBrokers.add(b);
+  private boolean rebalanceByMovingLoadIn(Broker broker,
+                                          ClusterModel clusterModel,
+                                          Set<Goal> optimizedGoals,
+                                          ActionType actionType,
+                                          Set<String> excludedTopics) {
+
+    if (!clusterModel.newBrokers().isEmpty() && !broker.isNew()) {
+      // We have new brokers and the current broker is not a new broker.
+      return true;
+    }
+
+    PriorityQueue<CandidateBroker> candidateBrokerPQ = new PriorityQueue<>();
+    // Sort the replicas initially to avoid sorting it every time.
+
+    double clusterUtilization = clusterModel.load().expectedUtilizationFor(resource()) / clusterModel.capacityFor(resource());
+    for (Broker candidate : clusterModel.healthyBrokers()) {
+      // Get candidate replicas on candidate broker to try moving load from -- sorted in the order of trial (descending load).
+      if (utilizationPercentage(candidate) > clusterUtilization) {
+        SortedSet<Replica> replicasToMoveIn = sortedCandidateReplicas(candidate, excludedTopics, 0, true);
```

Comment:
It is a little awkward that sometimes the boolean means isAscending and sometimes it means isDescending. Can we make them consistent?